Season config

mshafer1 opened the pull request:
creating a PR so I can make comments directly on the code.

harishbommakanti committed:
mostly finalized main page

-styled the nav menu btn and the side nav bar
- added instructions
- added github, twitter, and instagram pictures for the club. added club logo as tab image
- need to work on the season-config.html, event-data.html, and team-data.html pages now.

harishbommakanti committed:
added base code for  other pages

- added navbtn, title, and subtitle for other pages

harishbommakanti committed:
app is able to read newSeason input

- #navBtn --> #buttons to serve for all buttons
- basics of seasonConfig page are set up but need a way to handle all of the season names

harishbommakanti committed:
site is able to keep track of multiple seasons

- added seasonList and created methods to prevent duplicate seasons
- added a clearSeason button and added a "continue to event page" button
- need to work on passing the selected season to the event page

harishbommakanti committed:
only one radio button can be checked

- now, for the part when one checks a radio button, only one radio can be checked in total.
- need to work on storing the seasons on the page even after going back and forth between pages (loadSeasons() doesn't work, maybe cookies don't work when testing local html files)

harishbommakanti committed:
added some js, site still cannot remember data

- need some more code to manage cookies or another way to store data

harishbommakanti committed:
think the cookieList code is fixed

- think that the cookieList code is finished but need to get on a computer w/firefox to test

harishbommakanti committed:
COOKIE STORING FOR SEASONNAMES WORKS NOW

(title) - still looking into other methods like LocalStorage objects as this method isn't exactly the best over long term usage
- even if I stick to cookie storage I will need to alter the addNewSeason() and clearSeason() methods based on doc.cookie

harishbommakanti committed:
seasonNames cookies is now regular

- last few commits, doc.cookie was outputting irregular strings. now the doc.cookie is outputting a good semi-colon seperated value string
- need to work on the clear seasons button clearing all the cookies next.
- should also move the 'continue on to the event data' to the bottom page after questionnaire code

harishbommakanti committed:
commented out clearSeasons code

- decided that clearSeasons code isn't a priority (at least right now)
- will continue on to the questionnaire code and game specific elements code

harishbommakanti committed:
Start of question section for each season

- added charset tags
- added jquery download lilne to all files
- added a value attribute to seasonRadioButtons with  the seasonName
- added a button class in .css in case button had an id already
- added code to display a header for the current season checked

harishbommakanti committed:
started on forms for season specific elements

- added an onclick method for the new table row action
- added a table to structure the form around
- need to include the form and also deal with the data handling, preferably with cookies
- need to make it so that if i switch from season to season the appropriate elements pop up
- added some minor css stuff

harishbommakanti committed:
scoring table still not working, intermediate stage

- edited header
- need to work on html/js to see which elements (form, table etc) are easier to work with for scoring

harishbommakanti committed:
no input table anymore for season elements

- now, there is only 1 input box and prints all the elements in a table below
- now, need to work on storing the elements in doc.cookie, showing when the season name is checked only

harishbommakanti committed:
Element tables finalized

- fixed minor issue with trim() for adding seasons
- did away with the "check season" button so now only the radio button has to be checked to show the element table
- cannot enter the same element twice in table now
- have to refresh, toggle the button, and press the check season button to switch seasons BUT seeing as this is likely a 1 or two time thing it is not that critical.
- page remembers elements of different seasons
- now, work should be done on the questionnaire section for pit scouting

harishbommakanti committed:
starting commit for pit scouting area

- added textarea/css and other buttons for basic questionnaire layout
- will need to start parsing questions based on '.' or '?' and saving the data in cookies.
- will also need to add a confirmation box with messages like "please check a season"

harishbommakanti committed:
start on quest., refining prev scout config page

- added minor comments for code
- added some code for parsing the questions based on \n
- need to work on cookies not overwriting

- going through past seasonConfig code as per Matt's advice
- still need to work on new cookie format for elements
- need to work on parsing questions

harishbommakanti committed:
Refactored cookie naming

- minor changes from stdlib to jquery, some commenting, clearing input values/texts when stuff changes
- cookie naming changed and now, when the seasons are changed, the table is set with the elements stored in the cookies already
- deleted some unneeded commented code

harishbommakanti committed:
added keypress events

- keypress events for both the season input box and the scoringValue box, when enter is pressed
- now, should work on parsing the questions in the questionnaire box and adding it to cookies in a similar format than the seasons

harishbommakanti committed:
edge case where points=""

- added code for case where user doesn't enter points

harishbommakanti committed:
Questionnaire code finished

- minor .val("")'s added here and there to make sections blank when changing seasons
- can enter questionnaire data
- if you change to a different season, the page will load the saved questions into the text box, after clearing the current content

mshafer1 commented on home.html:
```diff
@@ -0,0 +1,58 @@
+<!DOCTYPE html>
```
this should be index.html as it is your main page

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+
+//adds new season from input box
+function addNewSeason()
+{
```
If you feel like you need section headers because the code is unrelated - you should move to separate files.

You can include multiple js files into one page - even with an ordering (the order you load them).

harishbommakanti replied:
I'll probably do this, there are like 300 lines for just the season config alone.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+        $("#seasonRadioList").append("<label><input type=\"radio\" name=\"seasonListItem\" value=\"" + newSeason + "\">" + newSeason + "</label><br>"); //add to radiolist
+
+        //save newSeason in cookie
+        document.cookie = "seasonList=" + getCookie("seasonList") + newSeason + "," + "; expires=Wed, 1 Jan 2038 12:00:00 UTC; path=/";
```
You're doing this a lot.

Clone this method from W3Schools (https://www.w3schools.com/js/js_cookies.asp):
```JS
function setCookie(cname, cvalue, exdays) {
  var d = new Date();
  d.setTime(d.getTime() + (exdays*24*60*60*1000));
  var expires = "expires="+ d.toUTCString();
  document.cookie = cname + "=" + cvalue + ";" + expires + ";path=/";
}
```

You could even tweak it to:
```JS
function setCookie(cname, cvalue, exdays=750) {
  var d = new Date();
  d.setTime(d.getTime() + (exdays*24*60*60*1000));
  var expires = "expires="+ d.toUTCString();
  document.cookie = cname + "=" + cvalue + ";" + expires + ";path=/";
}
'''
(notice the default on the endDays value).

Anywhere you find yourself writing the same or basically the same code more than twice, you should make a function (or object or VI . . . - but in JS, mostly functions).

harishbommakanti replied:
Yeah that was pretty useful, I had to change a lot of instances of document.cookie.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+    $("#elementKey").val("");
+    $("#elementPoints").val("");
+    $("#pitScoutConfirmationBox").html("");
+    $("#pitScoutingTextBox").val("");
```
give all of these elements a class named something like: `js_clear_on_load`
then this code turns into:

```JS
$('.js_clear_on_load').val("").html("");
```

much simpler, much easier to extend (just add the class to whatever else needs it).

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+    console.log("document.cookie: " + document.cookie);
+    var formSpace = document.getElementById("seasonRadioList");
+    formSpace.innerHTML = " ";
+}*/
```
Are you planning to use this code later? Like tomorrow? If not, delete it and use a commit to track what was there before.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+    var re = new RegExp(name + "=([^;]+)");
+    var value = re.exec(document.cookie);
+    return (value != null) ? unescape(value[1]) : " ";
+}
```
Good! Use helper methods for common operations.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+
+    $("#elementsTableTitle").text("Scoring table for season \'" + season + "\'");
+    
+    $("input[name=seasonListItem]:checked","#seasonRadioList").setCurrCheckedSeason();
```
`setCurrCheckedSeason` is this a standard JQuery method?

harishbommakanti replied:
It's the first method in the $(document).ready section

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+}
+
+//updates table based on cookies from checked season
+function setTable(seasonChecked)
```
at this point, it doesn't matter that it's the "checked season" - just that it's the season you care about. `season` or `season_to_load` would work better.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+        {
+            return currCheckedSeason;
+        }
+    }
```
Why is this twice?

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+
+    //if hits enter during season text box, trigger button
+    $('#newSeasonInputBox').keypress(function(e){
+        if (e.keyCode == 13)
```
What's 13?
You have a comment above, an easier to read option could be

```JS
const Enter_key_code = 13;
 $('#newSeasonInputBox').keypress(function(e){
        if (e.keyCode == Enter_key_code)
...
```

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+                } else
+                {
+                    $("#elementConfirmationBox").html('<i>Element "' + newElementName + '" already exists.<\i>');
+                }
```
What if I want to change it?

Probably just making it always over-write works better.

mshafer1 replied:
might also want to think of a way to delete - but a current way to do that would be to delete the cookie, so that's a lower priority.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+        if (typeof(currCheckedSeason) !== "undefined")
+        {
+            document.cookie = "currCheckedSeason="+currCheckedSeason+
+            "; expires=Wed, 1 Jan 2038 12:00:00 UTC; path=/";
```
use setCookie here, too.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+            }
+        }
+    }
+
```
If you're going to use the cookie to determine what season is selected each time you go to save (like at [line 291](https://github.com/Team6321/Scouting_App/pull/1/files#diff-9a9569e9d73f33740eada95275da7f30R291)), you should select this based on that cookie here as well - so the app doesn't behave like a season is selected when it doesn't look like one is.

That would look like (after you've loaded the curentSeason):

```JS
$(`input[name=seasonListItem][value="${ currentSeason }"]`).attr('checked', true);
```

`.attr(` gives access to just about any attribute on an element (text, html, value, checked) and is the under-the-hood code for most if not all of the easy accessors (`.text(`) - we use it when accessing (or setting) an attribute on an object that only that type has (hence, low return on investment in making an easy accessor for it).

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+
+    //adds new element to table, handles things like element already exists and season not checked
+    $("#scoringSaveButton").click(function(){
+        var newElementName = $("#elementKey").val().trim().toString();
```
isn't this value already a string ??

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+            var season = $("input[name=seasonListItem]:checked","#seasonRadioList").val();
+            if (typeof(season) !== "undefined")
+            {
+                var cookieName = '/season_config/' + season + '/' + newElementName;
```
you're doing this a lot - maybe make a helper method?

```JS
function season_config_cookie_name(season, element) {
  return `/season_config/${ season }/${ element }`;
}
```

This way, if you decide to tweak where the cookie is (say, `season_config` to `season_info`), but you used this helper method everywhere you either read or set a corresponding cookie, you can just change the function.

harishbommakanti replied:
That's neat, basically javascript's version of the java printf

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+                if (!elementAlreadyStored)
+                {
+                    $("#elementConfirmationBox").text(" ");
+                    document.cookie = cookieName + "=" + cookieValue + 
```
use `setCookie` here.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+                    $("#scoringTable").append(newTRHtml);
+                    $("#elementConfirmationBox").html('<i>Element "' + newElementName + '" added.<\i>');
+
+                    $("#elementKey").val("");
```
for ease of use, tweak this line to be:
```JS
$("#elementKey").val("").focus();
```

This will move the cursor back to the key field

This allows for:
```
Ball
TAB
5
Enter
Hatch
10
```

quicker input if I don't have to shift+tab back (or click)
You'll see this as well with Excel - start entering a row of items, enter 1, tab, enter a second, tab, enter a third, enter - it will take you to the next row at the column you entered the first item in - allow for quickly typing out a lot of data.

harishbommakanti replied:
That's a cool feature, thanks.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+
+    //start of questionnaire js code
+    $("#pitScoutingSubmit").click(function(){
+        var allQuestions = $("#pitScoutingTextBox").val().split('\n');
```
So, this will split on all OS's, but on Windows, at '\r' will be on the tail of each line. Are you trimming them? Otherwise, use the regex split I sent you to grab the '\r' as well if present.

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+                    if (!getCookie(cookieName).includes(question))
+                    {
+                        //cookie --> {season} questions = {q1},{q2},{q3}
+                        document.cookie = cookieName + '=' + getCookie(cookieName) + question + ',; expires=Wed, 1 Jan 2038 12:00:00 UTC; path=/';
```
A) use setCookie

B) - don't seperate on comma's. What if the user wants to use that in the question?
`How many cargo, if any, can you get in auto?` - this will get parsed back out at three separate.
separate on a character the user can't type. You can use `\n` (if you're trimming anyways), or a Unicode character (although, careful with this, I setup all my machines to let me type a set of Unicode characters for nice clean communication - like 15° or 220 Ω :) )

harishbommakanti replied:
A) ok
B) I didn't think of that, I ended up using Ω because most users won't be able to type that

mshafer1 commented on script.js:
```diff
@@ -6,4 +9,317 @@ function openNav(){
+                        if (!$("#pitScoutConfirmationBox").html().includes(confirmationText))
+                        {
+                            $("#pitScoutConfirmationBox").append(confirmationText);
+                        }
```
And, what if I want to clear one out?

For that matter, with reloading the text box to show already stored ones - this could get really annoying if I have 5 already and want to add one.

mshafer1 replied:
I see that if I clear the text box and add the one I want to add that I get back all 6, but what if I want to edit a line? current implementation is going to store both versions and reload them in order.

Should probably just make this store whatever is in the text box right now - that allows for updating, deleting, and adding without any major work (think of it like a text document - when the user hits save -they expect exactly that to be saved - not appended to the last version).

mshafer1 reviewed (commented):
Liking the usability progress - I made a deeper dive on this in terms of corner cases I think you should consider and keeping your code easy to extend/modify.

This is going to be a great tool. Thanks for working on it!

mshafer1 commented on styles.css:
```diff
@@ -1,26 +1,66 @@
+  border-radius: 12px;
+  font-size:16px;
+  margin: 4px 2px;
+}
```
Apply this to all buttons (or rather, all buttons with `id="buttons")?

This may be desirable right now, or even with buttons in general, but you should use the class setup
```CSS
.r3-button
{
  color: gold;
  background-color: maroon;
  border-color: black;
  border-radius: 12px;
  font-size:16px;
  margin: 4px 2px;
}
```
and then in the html
```html
<button class="r3-button">blah</button>
```

(I'm using r3 for r^3, like it's fine to do this for what you want the "standard" button on the site to be,

mshafer1 commented on season-config.html:
```diff
@@ -0,0 +1,86 @@
+        <link href="https://fonts.googleapis.com/css?family=Ubuntu&display=swap" rel="stylesheet">
+    </head>
+    <body id="background" onload="loadSeasons()">
+        <button style = "float:left"type="button" onclick = "openNav()" id="buttons">Navigation Menu</button>
```
If you use the `<button` element, you only need to specify the `type` if you're inside a `form` that's trying to assume the button is the submit.

harishbommakanti committed:
PR #1 Fixes

https://github.com/Team6321/Scouting_App/pull/1

- another feature to work on is the deletion of seasons/elements/questions via cookie deletion

mshafer1 commented on season-config.html:
```diff
@@ -0,0 +1,86 @@
+            </table>
+            <h3 id="elementConfirmationBox" class='js_clear_on_load'></h3> <br><br>
+
+        </fieldset>
```
I can't submit this anymore . . . did you intend to change away from form?

mshafer1 commented on seasonConfigScript.js:
```diff
@@ -0,0 +1,288 @@
+}
+
+//after onchange, show saved cookie questions of new season to the text box
+function setPitQuestionTextBox(season)
+{
+    $('#pitScoutingTextBox').val("");
+    var allQuestions = getCookie(season + ' pitQuestions').trim().split('Ω');
+    for (var i = 0; i < allQuestions.length;i++)
+    {
+        var question = allQuestions[i].trim();
+        var currText = $("#pitScoutingTextBox").val();
+        if (i != allQuestions.length-1)
+            $("#pitScoutingTextBox").val(currText + question + '\n');
+        else
+            $("#pitScoutingTextBox").val(currText + question); //no newline after last output line
+    }
+    
+    if ($("#pitScoutingTextBox").val() === '\n') //clearing newlines after onchange event
+    {
+        $("#pitScoutingTextBox").val("");
+    }
+}
```
this isn't getting called anymore

mshafer1 commented on seasonConfigScript.js:
```diff
@@ -0,0 +1,288 @@
+            }
+        }
+    }
+});
```
I can't get this to get called

mshafer1 commented on styles.css:
```diff
@@ -1,26 +1,56 @@
 
-#navBtn
+.r3-button
 {
-    color: gold;
-    background-color: maroon;
-    font-family: 'Ubuntu';
-    float:left;
-    display:block;
-    margin-left:auto;
-    margin-right:auto;
+  color: gold;
+  background-color: maroon;
+  border-color: black;
+  border-radius: 12px;
+  font-size:16px;
+  margin: 4px 2px;
+}
+
+.r3-button:hover {
+  color: #f1f1f1;
+}
+
+.r3-button:hover {
+  color: #f1f1f1;
+}
+
```
yes, much better. :)

mshafer1 reviewed (commented):
something got messed up, I can't use the page any more.

It does load the seasons into the selector, I can select a season, but I can't add scoring methods, and I can't save pit scouting questions. Tried to dig into why, got no errors on console, and breakpoints in corresponding click or save methods didn't get hit.

I suspect that in the moving, you lost where you registered for click events (or form submit).

harishbommakanti committed:
minor fix to 'PR #1 fixes' commit

- included functions in 'document.ready()' to try and fix problem of pitScouting and elementSubmit functions to work

harishbommakanti committed:
added delete seasons/elements buttons

- added buttons/functions that deletes cookies for seasons/season elements and refreshes page.
- Not needed for questionnaire as submitting the questions already overwrites cookie

harishbommakanti committed:
Deleting season improvements

- added § separator, character that most users won't type, better than previous ',' separator
- deleteAllSeasons --> deleteCheckedSeason, 1 click into 2 clicks to avoid misclick causing tragic mistake
- after deleting season, elements/pit questions are deleted too

mshafer1 reviewed (commented):
looking better!

App works pretty well.

There are some things we could talk about in terms of UX, but this is functional.

Ship it! (go ahead and merge into master, double check the merge went ok, then merge into gh-pages, test, and push).
